### PR TITLE
feat: new erc20 transfer tests

### LINKED ISSUE #151: improve `eth_call` tests to have more real-world scenarios
our `eth_call` tests are fairly basic right now, simply incrementing a counter contract

in the real world the use cases are usually a bit more complex, and have led to tests not being able to catch issues that come up post-deploy

at the very least, we should have tests for:
- basic ERC-20 functionality
- ERC-20 with a TransparentProxy wrapper (e.g. USDC token)

the proxy contract with it's `delegatecall` behavior has some edge cases that broke this behavior on mainnet today. tests can be the exact same between both with the exception of the `to` field

### PR DESCRIPTION
**Pull request by steph-rs**
Partly covers #151; new test environment from #160

I had to copy some code to new test env (`flashblocks-rpc/tests`).

Needs to be added to close #151 completely:
- [ ] ERC-20 with a TransparentProxy wrapper (e.g. USDC token)

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td><code>2</code> if repo is sensitive</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>1</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>2</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: crates/flashblocks-rpc/tests/erc20.rs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+        .to(ERC20_ADDRESS)
+        .gas_limit(100_000)
+        .input(encode_balance_of(account));
+
+    let result = provider.call(call.into()).await?;
+    Ok(parse_uint256(result))
+}
+
+#[tokio::test]
+async fn test_basic_eth_call_with_flashblocks() -> eyre::Result<()> {
```
Comment (haardikk21) [inline crates/flashblocks-rpc/tests/erc20.rs:178]: this test is ~the same as already existing in `rpc.rs`. Don't think it needs to exist in `erc20.rs`

File: crates/flashblocks-rpc/tests/erc20.rs [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+        .gas_limit(21_000);
+
+    let result = provider.call(call).await;
+    assert!(result.is_ok(), "eth_call should work with flashblocks");
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_erc20_balance_progression() -> eyre::Result<()> {
```
Comment (haardikk21) [inline crates/flashblocks-rpc/tests/erc20.rs:197]: all of these steps appear to be covered in the multi user balances test as well. 

im not sure what the `SimpleToken` contract is that you used, but it may be more interesting here to track balance changes across functions other than just `transfer` here to ensure correctness rather than this test which seems like a subset of `test_multi_user_balances`

Specifically, if a `mint` function is available - we should be able to track that. If a `burn` function is available, we should track that. And `transferFrom` (i.e. approve + transferFrom) would be great.

**Review by haardikk21 — CHANGES_REQUESTED** [review]
Generally looks good to me, some minor changes requested

**Comment by cb-heimdall** [comment]
Review Error for haardikk21 @ 2025-11-03 16:05:00 UTC
User failed mfa authentication, see [go/mfa-help](http://go/mfa-help)

**Comment by steph-rs** [comment]
after #166 should be reworked entirely, so closing
